docker/entrypoint.sh

```diff
@@ -0,0 +1,31 @@
+#!/bin/sh
+
+case "$3" in
+	Darwin*)	HOST_MACHINE=Mac;;
+	Linux*)		HOST_MACHINE=Linux;;
+	*)		HOST_MACHINE="Unknown:$3";;
+esac
+
+if [ "$HOST_MACHINE" = "Mac" ]; then
+	/bin/sh -c "$5"
+elif [ "$HOST_MACHINE" = "Linux" ]; then
+	USER=$(id -nu "$1" 2>&1)
+	GROUP=$(getent group "$2" | cut -d: -f1)
+
+	if [ -z "$GROUP" ]; then
+	    GROUP=logdna
+	    groupadd -g "$2" logdna
+	fi
+
+	if case "$USER" in *"no such user") true ;; *) false ;; esac; then
+	    USER=logdna
+	    useradd -g "$GROUP" -u "$1" -m logdna
+	fi
+
+	chown -R "$USER":"$GROUP" "$4"
+	runuser -u "$USER" -g "$GROUP" -- /bin/sh -c "$5"
```

Comment:
Why are we doing this at runtime of the image?

Reply:
I see what you are doing.. i'm still at a loss as to why you need to do this :/

Reply:
I am trying to ensure two things:
1. The guest machine should have permissions to modify existing files; needed for `make clean`.
2. The host machine has correct ownership of files created by the guest machine; byproduct of `make build`, `make test`, and `make lint`.

Docker on the Mac platform gives us this ability out of the box:
1. The guest machine runs as root
2. Despite writing as root on the guest machine, Mac's filesystem correctly writes the files on the host machine with the permissions of the calling user/group.

Linux; however, does not. The filesystem will copy the permissions from the guest machine to the host machine so if the guest machine is running as root then the files are created as root on the host machine. To get around this, we need to run the guest machine as the user on the guest machine.

We can't simply pass the user into the docker run command because Docker always mounts the host directory as `1000:1000` so if the host machine user/group does not share this id it won't have access. The entrypoint script allows us to start the machine as root, chown the mounted directory to the correct user/group, and then run as that user/group with `runuser`.

Finally `runuser` won't work with ids so we need to ensure a user and group exists with the passed in ids. If they don't we need to create them.

This is ultimately to allow commands like `make build`, `make clean`, and `make test` to work both for developers (Mac & Linux) and our CI/CD system (Linux).

Reply:
I'm pretty sure it mounts the directory with the permissions the directory on the host machine is created as.
if the first step is `rm -r target && mkdir -p target` I think the directory permissions will be preserved.

This all might be a bit of over kill
I personally would just run `cargo build` 🤷 or  chown the artifact this myself anyway. 
But maybe you don't need a container for this. can you just use a shell script on the host and chown it as the user?

If the goal is to provide an artifact so people don't have to build it, could we just host pre-built binaries on jenkins or github?

Reply:
Second question - Does this build actually work on Mac? Can you build a rust binary in a linux container and Have it run on a mac laptop?

Reply:
@esatterwhite Ah while that seems to be true for Linux, it's not for Mac. We could change the script to run on the host and issue the correct docker run command:
```console
# Mac:
docker run --rm -w /build -v /Users/jake/logdna/logdna-agent-v2:/build:Z docker.io/rust:1.42 make build

# Linux:
docker run --rm -u $(id -u):$(id -g) -w /build -v /Users/jake/logdna/logdna-agent-v2:/build:Z docker.io/rust:1.42 make build
```

The purpose for using Docker, afaik, is to allow anyone to come along and build/test/lint the agent without needing to install cargo. This also ensures everyone is running on the same version and on the same platform to remove any inconsistencies between environments.

The build artifacts (the final binary as well as the compilation units) are, afaik, not compatible with Mac. The agent can't be built against Mac either because it relies on `inotify` which Mac doesn't provide. The benefits of keeping the build artifacts (namely the compilation units) is the ability to re use them for future builds/tests/lints.

Reply:
Rustup solves the version probably doesn't it?
you might want to look at the [notify](https://github.com/notify-rs/notify/tree/v4.0.15#notify) crate.